crates/wasmtime/src/engine.rs

```diff
@@ -68,6 +165,77 @@ impl Engine {
     pub fn same(a: &Engine, b: &Engine) -> bool {
         Arc::ptr_eq(&a.inner, &b.inner)
     }
+
+    /// Defines a function for the [`Engine`] for the given callback.
+    ///
+    /// Use [`Store::get_engine_func`] to get a [`Func`] representing the function.
+    ///
+    /// Note that the implementation of `func` must adhere to the `ty`
+    /// signature given, error or traps may occur if it does not respect the
+    /// `ty` signature.
+    ///
+    /// Additionally note that this is quite a dynamic function since signatures
+    /// are not statically known. For performance reasons, it's recommended
+    /// to use [`Engine::wrap_func`] if you can because with statically known
+    /// signatures the engine can optimize the implementation much more.
+    pub fn define_func(
+        &self,
+        module: &str,
+        name: &str,
+        ty: FuncType,
+        func: impl Fn(Caller<'_>, &[Val], &mut [Val]) -> Result<(), Trap> + Send + Sync + 'static,
+    ) -> Result<()> {
+        let mut funcs = self.inner.funcs.write().unwrap();
+
+        let ty_clone = ty.clone();
+
+        // Create a trampoline that converts raw u128 values to `Val`
+        let func = Box::new(move |caller_vmctx, values_vec: *mut u128| {
+            // Lookup the last registered store as engine functions have no associated store
+            let store = wasmtime_runtime::with_last_info(|last| {
+                last.and_then(Any::downcast_ref::<Store>)
+                    .cloned()
+                    .expect("Engine function called without thread state")
+            });
+
+            Func::invoke(&store, &ty_clone, caller_vmctx, values_vec, &func)
+        });
+
+        let (instance, trampoline) =
+            crate::trampoline::create_engine_function(&ty, func, &self.inner.config)?;
+
+        funcs.insert(
+            module,
+            name,
+            EngineFunc {
+                ty,
+                instance,
+                trampoline,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Defines a function for the [`Engine`] from the given Rust closure.
+    ///
+    /// Use [`Store::get_engine_func`] to get a [`Func`] representing the function.
+    ///
+    /// See [`Func::wrap`] for information about accepted parameter and result types for the closure.
+    pub fn wrap_func<Params, Results>(
+        &self,
+        module: &str,
+        name: &str,
+        func: impl IntoFunc<Params, Results> + Send + Sync,
+    ) {
```

Comment:
Since `define_func` above returns a `Result<()>` should this as well?

Reply:
`Func::new` returns `Self` because of the `expect` call on the result from `generate_func_export`.

Should we just change `define_func` to return `()` similarly to match?

Reply:
Yeah that seems fine to me! We'll probably want to document the shadowing behavior though since this differs slightly than that of `Linker` where by default `Linker` returns an error. Returning an error here though feels a bit heavyweight.

Reply:
A non-issue if these get removed from `Engine` in favor of the builder-pattern, though.

Re: shadowing, this PR should include a check in `Linker` that ensures failure if an engine function exists and the linker is `!allow_shadowing`.  I'll fix that.